File: Templates/backtracking.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        answer = []
        
        def dfs(start: int, current_nums: list[int]) -> None:
            nonlocal answer

            answer.append(current_nums.copy())

            for i in range(start, len(nums)):
                current_nums.append(nums[i])
                dfs(i + 1, current_nums)
                current_nums.pop()

        dfs(0, [])

        return answer

File: Templates/backtracking.cpp
#include <vector>
#include <functional>

class Solution {
public:
    std::vector<std::vector<int>> subsets(std::vector<int>& nums) {
        std::vector<std::vector<int>> answer;
        std::vector<int> current;
        std::function<void(int)> dfs = [&](int start) {
            answer.push_back(current);
            for (int i = start; i < (int)nums.size(); ++i) {
                current.push_back(nums[i]);
                dfs(i + 1);
                current.pop_back();
            }
        };
        dfs(0);
        return answer;
    }
};
